pkg/persistentvolumeclaim/debug_cstor.go

```diff
@@ -0,0 +1,185 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolumeclaim
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/fatih/color"
+
+	cstortypes "github.com/openebs/api/v2/pkg/apis/types"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+)
+
+// DebugCstorVolumeClaim is used to debug a cstor volume by calling various modules
+func DebugCstorVolumeClaim(k *client.K8sClient, pvc *corev1.PersistentVolumeClaim, pv *corev1.PersistentVolume) error {
+	// 1. Main Struture Creation which contains all cstor CRs, this structure will be passed accross all modules.
+	var cstorResources util.CstorVolumeResources
+	cstorResources.PVC, _ = k.GetPVC(pvc.Name, "default")
+	cstorResources.PV = pv
+	// 2. Fill in the available CRs
+	if pv != nil {
+		cv, _ := k.GetCV(pv.Name)
+		cstorResources.CV = cv
+		cvc, _ := k.GetCVC(pv.Name)
+		cstorResources.CVC = cvc
+		cva, _ := k.GetCVA(util.CVAVolnameKey + "=" + pv.Name)
+		cstorResources.CVA = cva
+		cvrs, _ := k.GetCVRs(cstortypes.PersistentVolumeLabelKey + "=" + pv.Name)
+		cstorResources.CVRs = cvrs
+	}
+	sc, _ := k.GetSC(*pvc.Spec.StorageClassName)
+	// 3. Fill in the Pool and Blockdevice Details
+	if sc != nil {
+		cspc, _ := k.GetCSPC(sc.Parameters["cstorPoolCluster"])
+		cstorResources.CSPC = cspc
+		if cspc != nil {
+			cspis, _ := k.GetCSPIs(nil, "openebs.io/cas-type=cstor,openebs.io/cstor-pool-cluster="+cspc.Name)
+			cstorResources.CSPIs = cspis
+			expectedBlockDevicesInPool := make(map[string]bool)
+			// This map contains the list of BDs we specified at the time of Pool Creation
+			for _, pool := range cspc.Spec.Pools {
+				dataRaidGroups := pool.DataRaidGroups
+				for _, dataRaidGroup := range dataRaidGroups {
+					for _, bdName := range dataRaidGroup.GetBlockDevices() {
+						expectedBlockDevicesInPool[bdName] = false
+					}
+				}
+			}
+			// This list contains the list of BDs which are actually present in the system.
+			var presentBlockDevicesInPool []string
+			for _, pool := range cspis.Items {
+				raidGroupsInPool := pool.GetAllRaidGroups()
+				for _, item := range raidGroupsInPool {
+					presentBlockDevicesInPool = append(presentBlockDevicesInPool, item.GetBlockDevices()...)
+				}
+			}
+
+			// Mark the present BDs are true.
+			cstorResources.PresentBDs, _ = k.GetBDs(presentBlockDevicesInPool, "")
+			for _, item := range cstorResources.PresentBDs.Items {
+				if _, ok := expectedBlockDevicesInPool[item.Name]; ok {
```

Comment:
[TODO Later] Need to think & discuss if they should be only set to true if that BlockDevice is in Active state(not claimState).

The BD can go in Inactive/Unknown state depending on various factors such as disk detached, NDM daemonset pod in CLBO state on that node.

Reply:
This, shows which are present and which are not, if anything is present its status would be shown otherwise, it would be shown as not found.